### PR TITLE
[WIP] Fix current_epoch value on training end

### LINKED ISSUE #5007: Duplicate epochs when calling .fit() twice
### To Reproduce

```python
def test_bug(tmpdir):
    epochs = []
    
    class TestModel(BoringModel):
        def on_epoch_end(self):
            epochs.append(self.current_epoch)
    
    trainer = Trainer(
        max_epochs=2,
        limit_train_batches=1,
        limit_val_batches=1,
        default_root_dir=tmpdir,
        checkpoint_callback=False,
        logger=False,
        weights_summary=None,
        progress_bar_refresh_rate=0,
    )
    trainer.fit(TestModel())
    trainer.max_epochs=4
    trainer.fit(TestModel())

    assert epochs == list(range(4))
    # AssertionError [0, 1, 1, 2, 3] != [0, 1, 2, 3]
```

### Expected behavior

Assertion does not fail

### Environment

Current master

cc @tchaton @borda

### PR DESCRIPTION
**Pull request by carmocca**
# Blocked by #8477 and enabling restoring the ckpt progress tracking state by default.

## What does this PR do?

Fixes #5007
Part of #7406 

### Does your PR introduce any breaking changes? If yes, please list them.

With this PR, a `Trainer(max_epochs=1, limit_train_batches=1)` which saves a checkpoint`on_train_epoch_end` will have the `current_epoch=0, global_step=1` values saved. This is because we consider `on_train_epoch_end` to be part of the epoch itself:

```python
for epoch in epochs:
  ...
  on_train_epoch_end()
  epoch_count++
on_train_end()
```

`current_epoch` is now increased by 1 when training is over.
This means that if a model is run for 3 epochs (0, 1, 2), `trainer.current_epoch` will return `3`.
This is consistent with the fact that a new trainer returns `trainer.current_epoch == 0`, as in, the 0-th (first) epoch needs to run.
This will be breaking for anybody accessing the `trainer.current_epoch` attribute after `fit` is finished

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

🤪

### CODE REVIEW

File: tests/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -951,30 +955,28 @@ def assert_checkpoint_log_dir(idx):
...
     trainer.test(model)
-    assert trainer.current_epoch == epochs - 1
+    assert trainer.current_epoch == epochs
 
     for idx in range(1, 5):
         chk = get_last_checkpoint(ckpt_dir)
         assert_checkpoint_content(ckpt_dir)
 
         # load from checkpoint
-        trainer_config["callbacks"] = [ModelCheckpoint(dirpath=ckpt_dir, save_top_k=-1)]
```
Comment (carmocca) [inline tests/checkpointing/test_model_checkpoint.py:969]: Need to remove this because creating a new instance means we lose the existing value for

`ModelCheckpoint._last_global_step_saved`

which avoids saving a duplicate checkpoint on the second fit call. Perhaps this value should be saved and reloaded so that creating a new instance also works.

File: tests/callbacks/test_timer.py [hunk 2]
```diff
@@ -126,14 +126,8 @@ def test_timer_zero_duration_stop(tmpdir, interval):
     timer = Timer(duration=duration, interval=interval)
     trainer = Trainer(default_root_dir=tmpdir, callbacks=[timer])
     trainer.fit(model)
-    if interval == "step":
-        # timer triggers stop on step end
-        assert trainer.global_step == 1
-        assert trainer.current_epoch == 0
-    else:
-        # timer triggers stop on epoch end
-        assert trainer.global_step == len(trainer.train_dataloader)
```
Comment (carmocca) [inline tests/callbacks/test_timer.py:135]: This is no longer the case because min_epochs is now 0 so this check resolves to True and we actually stop at the beginning.

https://github.com/PyTorchLightning/pytorch-lightning/blob/a6a28e08d22f59f2468ff1049c67507df7519e7c/pytorch_lightning/loops/fit_loop.py#L160

File: pytorch_lightning/loops/fit_loop.py [hunk 3]
```diff
@@ -129,6 +119,18 @@ def running_loss(self) -> TensorRunningAccum:
         """Returns the running loss."""
         return self.epoch_loop.batch_loop.running_loss
 
+    @Loop.restarting.setter
+    def restarting(self, restarting: bool) -> None:
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:123]: Note to self: move this to `load_state_dict`?

File: tests/models/test_hooks.py [hunk 4]
```diff
@@ -611,16 +613,15 @@ def test_trainer_model_hook_system_fit_no_val_and_resume(tmpdir):
...
         "global_step": 1,
         "lr_schedulers": ANY,
         "optimizer_states": ANY,
         "pytorch-lightning_version": __version__,
         "state_dict": ANY,
         "loops": ANY,
     }
-    # TODO: wrong saved epoch, should be 0
-    saved_ckpt = {**loaded_ckpt, "global_step": steps_after_reload, "epoch": 2}
+    saved_ckpt = {**loaded_ckpt, "global_step": steps_after_reload, "epoch": 1}
```
Comment (carmocca) [inline tests/models/test_hooks.py:624]: Epoch is still 1 because the test specifies `max_steps=1` which counts as 1 epoch run (of 1 step).

**Review by awaelchli — COMMENTED** [review]
I think one of the most important things to remember from this PR is the fact that `on_train_epoch_end` checkpoints are different from the regular checkpoint since they have a different "fake epoch" (current epoch) value saved.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -336,21 +335,12 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
...
-        global_step += 1
-        if not has_reached_max_steps:
-            current_epoch += 1
-
         model = self.trainer.lightning_module
 
         checkpoint = {
-            "epoch": current_epoch,
-            "global_step": global_step,
+            # the epoch is saved for compatibility but it's not relevant for restoration
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:341]: Global step is currently not managed by progress tracking. In the future it will.

File: tests/checkpointing/test_model_checkpoint.py [hunk 6]
```diff
@@ -951,15 +955,15 @@ def assert_checkpoint_log_dir(idx):
...
     trainer.validate(model)
-    assert trainer.current_epoch == epochs - 1
+    assert trainer.current_epoch == epochs
 
     trainer.test(model)
-    assert trainer.current_epoch == epochs - 1
+    assert trainer.current_epoch == epochs
 
     for idx in range(1, 5):
         chk = get_last_checkpoint(ckpt_dir)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/checkpointing/test_model_checkpoint.py:969]: Yeah, we have several unnecessarily complicated checkpoint tests. But I don't want to touch them and silently open the door to regressions.

File: tests/callbacks/test_early_stopping.py [hunk 7]
```diff
@@ -260,7 +260,7 @@ def validation_epoch_end(self, outputs):
         max_epochs=10,
     )
     trainer.fit(model)
-    assert trainer.current_epoch == expected_stop_epoch
+    assert trainer.current_epoch - 1 == expected_stop_epoch
```
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/callbacks/test_early_stopping.py:263]: So this test has `expected_stop_epoch=2` which is used to index a list. Thus we know that this is a 0-based index. 2 idx == 3 count.

This means that 3 epochs will run and then training will stop.

The behaviour doesn't change between master and this PR. What changed is that we `completed++` now in `FitLoop.on_run_end` because this removed `        self.current_epoch = max(self.current_epoch - 1, 0)`

[the remark this answers is not in the document]
Comment (carmocca) [reply tests/callbacks/test_early_stopping.py:263]: `expected_stop_epoch` is confusing because it should be named `expected_stop_epoch_index` by the way it's used.

File: pytorch_lightning/loops/fit_loop.py [hunk 8]
```diff
@@ -142,31 +157,23 @@ def done(self) -> bool:
...
-        if self.trainer.should_stop:
-            # early stopping
-            met_min_epochs = self.epoch_progress.current.completed >= self.min_epochs if self.min_epochs else True
-            met_min_steps = self.global_step >= self.min_steps if self.min_steps else True
-            if met_min_epochs and met_min_steps:
-                should_stop = True
-            else:
+        stop_epochs = _is_max_limit_reached(self.epoch_progress.current.processed, self.max_epochs)
+
+        if self.trainer.should_stop and self.min_epochs:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:162]: I understand that is now a limitation. But is it a bug of a feature?

For proper loop customization we need to respect the hierarchy imposed by the loops. The way they are, epochs go before steps so they should always be prioritized when both are set.

[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/fit_loop.py:162]: > The way they are, epochs go before steps so they should always be prioritized when both are set.

I disagree. There is no reason to believe that "epochs" are more important than "steps". A setting like `Trainer(min_epochs=1, min_steps=50)` would be contradictory if we stopped at 1 epoch without reaching 50 steps. The arguments in the Trainer mean, if I request `Trainer(x=A, y=B, z=C)` I want features A AND B AND C together, not A OR B OR C.

File: pytorch_lightning/loops/fit_loop.py [hunk 9]
```diff
@@ -185,23 +189,6 @@ def on_run_start(self) -> None:  # type: ignore[override]
         # reset train dataloader and val dataloader
         self.trainer.reset_train_val_dataloaders(self.trainer.lightning_module)
 
-        ft_enabled = _FaultTolerantMode.detect_current_mode().is_enabled
```
Comment (rohitgr7) [inline pytorch_lightning/loops/fit_loop.py:188]: quick ques: we are not raising this warning anymore even if ckpt is mid epoch but no fault-tolerant is enabled?

Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:187]: so is it fault-tolerant by default?

Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:187]: The loop progress restoration yes.

The dataloading, rng, etc no.

File: pytorch_lightning/loops/fit_loop.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:162]: I don't think we can do that, because the intersection depends on the structure of the loops.

When `fit_loop.done is True` we would need to ask the children loops, but to do that they need to run their `reset` hook which is only called in their `run`.

We cannot know if the user wants the intersection or the union.

File: pytorch_lightning/loops/fit_loop.py (hunk 9 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/fit_loop.py:187]: if dataloader isn't reset and is going to iterate over it again completely, to me it looks more like a fresh epoch start, even though the progress is intact from the previous run.

Comment (carmocca) [reply pytorch_lightning/loops/fit_loop.py:187]: I looked further into this. I added back this warning and it's tests. But I had to move it one loop down.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 10]
```diff
@@ -118,6 +119,16 @@ def reset(self) -> None:
...
             self.batch_loop.optimizer_loop.optim_progress.reset_on_restart()
+
+            trainer = self.trainer
+            if not trainer.state._fault_tolerant_mode.is_enabled and trainer.num_training_batches != float("inf"):
+                expected_steps = math.ceil(trainer.num_training_batches / trainer.accumulate_grad_batches)
+                if self.global_step % expected_steps != 0:
+                    rank_zero_warn(
+                        "You're resuming from a checkpoint that ended before the epoch ended. This can cause unreliable"
+                        "results if further training is done. Consider using an end-of-epoch checkpoint or enabling"
+                        "fault-tolerant training."
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:130]:
```suggestion
                        "You're resuming from a checkpoint that ended before the epoch ended. This can cause unreliable"
                        " results if further training is done. Consider using an end-of-epoch checkpoint or enabling"
                        " fault-tolerant training."
```

**Review by awaelchli — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Comment by rohitgr7** [comment]
just one quick ques:

checkpoints saved during checkpointing... let's say during the last epoch, be it inside `on_validation_end` or `on_train_epoch_end`, the epoch values saved and reloaded will be different from what will be saved if `trainer.save_checkpoint` is called after training?

**Comment by carmocca** [comment]
Correct. There are several tests testing both cases.

**Comment by carmocca** [comment]
Yes.

I'm looking at global step in #11805 (don't look inside yet!)

File: pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 11]
```diff
@@ -221,13 +221,13 @@ def on_train_epoch_end(self, trainer: "pl.Trainer", *args):
         trainer.fit_loop._skip_backward = False
 
     def on_train_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
-        if self._model_contains_batch_norm and trainer.current_epoch == self.swa_end + 1:
+        if self._model_contains_batch_norm and trainer.current_epoch - 1 == self.swa_end + 1:
```
Comment (tchaton) [inline pytorch_lightning/callbacks/stochastic_weight_avg.py:224]: Add a comment here current_epoch is increased by 1.

File: pytorch_lightning/loops/fit_loop.py [hunk 12]
```diff
@@ -144,12 +157,12 @@ def done(self) -> bool:
         """Evaluates when to leave the loop."""
         # TODO(@awaelchli): Move track steps inside training loop and move part of these condition inside training loop
         stop_steps = _is_max_limit_reached(self.global_step, self.max_steps)
-        stop_epochs = _is_max_limit_reached(self.epoch_progress.current.completed, self.max_epochs)
+        stop_epochs = _is_max_limit_reached(self.epoch_progress.current.processed, self.max_epochs)
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:160]: Mind explaining the change?

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 13]
```diff
@@ -118,6 +119,16 @@ def reset(self) -> None:
...
             self.batch_loop.optimizer_loop.optim_progress.reset_on_restart()
+
+            trainer = self.trainer
+            if not trainer.state._fault_tolerant_mode.is_enabled and trainer.num_training_batches != float("inf"):
+                expected_steps = math.ceil(trainer.num_training_batches / trainer.accumulate_grad_batches)
+                if self.global_step % expected_steps != 0:
+                    rank_zero_warn(
+                        "You're resuming from a checkpoint that ended before the epoch ended. This can cause unreliable"
+                        " results if further training is done. Consider using an end-of-epoch checkpoint or enabling"
+                        " fault-tolerant training."
```
Comment (tchaton) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:130]: Mind adding a link to the Fault Tolerant documentation ?

File: pytorch_lightning/loops/fit_loop.py [hunk 14]
```diff
@@ -309,11 +305,6 @@ def on_advance_end(self) -> None:
     def on_run_end(self) -> None:
         """Calls the ``on_train_end`` hook."""
         log.detail(f"{self.__class__.__name__}: train run ended")
-        # NOTE: the current_epoch is already incremented
-        # Lightning today does not increment the current epoch at the last epoch run in Trainer.fit
-        # To simulate that current behavior, we decrement here.
-        # TODO: must be fixed by https://github.com/PyTorchLightning/pytorch-lightning/issues/5007
-        self.epoch_progress.current.completed = max(self.epoch_progress.current.completed - 1, 0)
```
Comment (tchaton) [inline pytorch_lightning/loops/fit_loop.py:316]: Neat !

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 15]
```diff
@@ -327,21 +326,12 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
...
+            "epoch": self.trainer.current_epoch,
+            "global_step": self.trainer.global_step + 1,
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:334]: Should we increment global_step there? Normally, it shouldn't be if a failure was raised. Am I missing something ?

File: pytorch_lightning/tuner/batch_size_scaling.py [hunk 16]
```diff
@@ -60,10 +60,8 @@ def scale_batch_size(
 
     # Save initial model, that is loaded after batch size is found
     ckpt_path = os.path.join(trainer.default_root_dir, f".scale_batch_size_{uuid.uuid4()}.ckpt")
-    trainer.fit_loop.epoch_progress.current.completed -= 1
```
Comment (tchaton) [inline pytorch_lightning/tuner/batch_size_scaling.py:63]: Super neat !

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 15 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:334]: That's how it works in master, it gets increased unconditionally:

https://github.com/PyTorchLightning/pytorch-lightning/blob/5e78f429e3a79ac2c7c7402f922c179775db88a4/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L338

Preparing a fix (after this PR) in #11805

File: pytorch_lightning/loops/base.py [hunk 17]
```diff
@@ -288,11 +287,9 @@ def state_dict(self, destination: Optional[Dict] = None, prefix: str = "") -> Di
 
         destination[prefix + "state_dict"] = self.on_save_checkpoint()
 
-        # do not get the mode from `self.trainer` because it might not have been attached yet
-        ft_enabled = _FaultTolerantMode.detect_current_mode().is_enabled
         for k, v in self.__dict__.items():
             key = prefix + k
-            if ft_enabled and isinstance(v, BaseProgress):
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/loops/base.py:295]: It's necessary for this PR

**Review by tchaton — APPROVED** [review]
LGTM !
